content/frameworks/react/virtualize-long-lists-react-window/index.md

```diff
@@ -0,0 +1,286 @@
+);
+
+const getItemSize = index => {
+  // return a size for items[index]
+}
+ 
+const ListComponent = () => (
+  <VariableSizeList
+    height={500}
+    width={500}
+    itemCount={items.length}
+    itemSize={getItemSize}
+  >
+    {Row}
+  </VariableSizeList>
+);
+
+export default ListComponent;
+```
+
+The following embed shows an example of this component.
+
+<div class="glitch-embed-wrap" style="height: 750px; width: 100%;">
+  <iframe
+    allow="geolocation; microphone; camera; midi; vr; encrypted-media"
+    src="https://glitch.com/embed/#!/embed/react-window-variable-list?path=src/ListComponent.js&previewSize=100"
+    alt="react-window-variable-list on Glitch"
+    style="height: 100%; width: 100%; border: 0;">
+  </iframe>
+</div>
+
+The item size function passed to `itemSize` prop randomizes the row heights in this example. In a real application however, there should be actual logic defining the sizes of each item.
+
+<aside class="aside note">
+Both <code>FixedSizeList</code> and <code>VariableSizeList</code> components support horizontal lists by using a <code>layout="horizontal"</code> prop. Take a look at the <a href="https://react-window.now.sh/#/examples/list/fixed-size">documentation</a> to see an example.
+</aside>
+
+### Grids
+
+`react-window` also provides support for virtualizing multi-dimensional lists, or grids. In this context, the "window" of visible content changes as the user scrolls horizontally **and** vertically.
+
+![Window diagram for a Grid](./window-diagram-grid.png)
+
+Similary, both `FixedSizeGrid` and `VariableSizeGrid` components can be used depending on whether the size of specific list items can vary. 
+
+* For `FixedSizeGrid`, the API is about the same but with the fact that heights, widths and item counts need to be represented for both columns and rows.
+* For `VariableSizeGrid`, both the column widths and row heights can be changed by passing in functions instead of values to their respective props.
+
+Take a look at the [documentation](https://react-window.now.sh/#/examples/grid/fixed-size) to see examples of virtualized grids.
+
+<aside class="aside note">
+  Aside from providing the base components to create efficient lists and grids, <code>react-window</code> also provides other capabilities such as scrolling to a specific item or providing an indicator when the user is scrolling. The <a href="https://react-window.now.sh/#/examples/list/scrolling-indicators">documentation</a> provides examples for this.
+</aside>
+
+## Lazy loading on scroll
+
+Many websites take advantage of not loading all the items on a list if further rows need to be obtained over a network call. This technique, commonly referred to as "infinte loading", adds new DOM nodes into the list as the user scrolls past a certain threshold close to the end. Although this is better than loading all items on a list at once, it still ends up populating the DOM with thousands of row entries if the user has scrolled past that many.
```

Comment:
My attempt to simplify the first sentence:  "Many websites improve performance by waiting to load and render items in a long list until the user has scrolled down."

Also, it feels like we could add a last little bit here to explain why thousands of entries is a problem:

"This can lead to an excessively large DOM size, which starts to impact performance by making style calculations and DOM mutations slower."

Reply:
Added 👍